packages/alfa-rules/src/helpers/predicate-builder.ts

```diff
@@ -0,0 +1,160 @@
+import { getRole, Role } from "@siteimprove/alfa-aria";
+import { BrowserSpecific } from "@siteimprove/alfa-compatibility";
+import { Device } from "@siteimprove/alfa-device";
+import {
+  Element,
+  getElementNamespace,
+  getInputType,
+  InputType,
+  isElement as isElt,
+  Namespace,
+  Node
+} from "@siteimprove/alfa-dom";
+import { Predicate } from "@siteimprove/alfa-util";
+
+const {
+  BinOp: { and, or }
+} = BrowserSpecific;
+
+class PredicateBuilder<T, U extends T = T> {
+  public readonly predicate: Predicate<T, U>;
+
+  public constructor(predicate: Predicate<T, U> = () => true) {
+    this.predicate = predicate;
+  }
+
+  public and<V extends U = U>(
+    predicate: Predicate<U, V>
+  ): PredicateBuilder<T, V> {
+    return new PredicateBuilder(v => this.predicate(v) && predicate(v));
+  }
+
+  public browserSpecific(): BrowserSpecificPredicateBuilder<T, U> {
+    return new BrowserSpecificPredicateBuilder(this.predicate);
+  }
+}
+
+class NodePredicateBuilder<U extends Node = Node> extends PredicateBuilder<
+  Node,
+  U
+> {
+  public isElement(): ElementPredicateBuilder {
+    return new ElementPredicateBuilder(node => {
+      return this.and(isElt).predicate(node);
+    });
+  }
+}
+
+class ElementPredicateBuilder<
+  U extends Element = Element
+> extends NodePredicateBuilder<U> {
+  public and<V extends U = U>(
+    predicate: Predicate<U, V>
+  ): ElementPredicateBuilder<V> {
+    return new ElementPredicateBuilder<V>(
+      element => this.predicate(element) && predicate(element)
+    );
+  }
+
+  public withName(...names: Array<string>): ElementPredicateBuilder {
+    return this.and(element => names.some(name => name === element.localName));
+  }
+
+  public withInputType(
+    ...inputTypes: Array<InputType>
+  ): ElementPredicateBuilder {
+    return this.and(element =>
+      inputTypes.some(inputType => inputType === getInputType(element))
+    );
+  }
+
+  public withNamespace(
+    context: Node,
+    ...namespaces: Array<Namespace>
+  ): ElementPredicateBuilder {
+    return this.and(element =>
+      namespaces.some(
+        namespace => namespace === getElementNamespace(element, context)
+      )
+    );
+  }
+
+  public withRole(
+    device: Device,
+    context: Node,
+    ...roles: Array<Role>
+  ): BrowserSpecificPredicateBuilder<Node, U> {
+    return this.browserSpecific().and(element =>
+      BrowserSpecific.map(getRole(element, context, device), elementRole =>
+        roles.some(role => role === elementRole)
+      )
+    );
+  }
+}
+
+type BrowserSpecificPredicate<T, U extends T = T> = (
+  t: T
+) => boolean | BrowserSpecific<boolean>; // t is U
+
+class BrowserSpecificPredicateBuilder<T, U extends T = T> {

```

Comment:
This should somehow inherit from `PredicateBuilder` because now `withName` and such must be used before `withRole` and duplicating them would be bad. Not sure this is really possible 😞